### ISSUE TITLE
NextLS crashing on project load

### ISSUE THREAD

**Issue by John-Goff** [elixir-tools/next-ls#335]
Immediately upon loading LSP on any files in any projects that I've tried, I get a crash with the following messages in the log buffer.

```
[NextLS] Booting runtimes...
LSP Exited.

Last message received: handle_notification %{"jsonrpc" => "2.0", "method" => "initialized", "params" => %{}}

** (Protocol.UndefinedError) protocol Enumerable not implemented for nil of type Atom. This protocol is implemented for the following type(s): DBConnection.PrepareStream, DBConnection.Stream, Date.Range, Exqlite.Stream, File.Stream, Function, GenEvent.Stream, HashDict, HashSet, IO.Stream, Jason.OrderedObject, List, Map, MapSet, Range, Stream
    (elixir 1.15.7) lib/enum.ex:1: Enumerable.impl_for!/1
    (elixir 1.15.7) lib/enum.ex:166: Enumerable.reduce/3
    (elixir 1.15.7) lib/enum.ex:4387: Enum.reduce/3
    (next_ls 0.15.0) lib/next_ls.ex:643: NextLS.handle_notification/2
    (gen_lsp 0.6.2) lib/gen_lsp.ex:285: anonymous fn/4 in GenLSP.loop/3
    (gen_lsp 0.6.2) lib/gen_lsp.ex:308: GenLSP.attempt/3
    (stdlib 5.1.1) proc_lib.erl:241: :proc_lib.init_p_do_apply/3



LSP Exited.

Last message received: handle_request %{"id" => 1909, "jsonrpc" => "2.0", "method" => "initialize", "params" => %{"capabilities" => %{"general" => %{"positionEncodings" => ["utf-32", "utf-16"]}, "textDocument" => %{"callHierarchy" => %{"dynamicRegistration" => false}, "codeAction" => %{"codeActionLiteralSupport" => %{"codeActionKind" => %{"valueSet" => ["", "quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor.rewrite", "source", "source.organizeImports"]}}, "dataSupport" => true, "dynamicRegistration" => true, "isPreferredSupport" => true, "resolveSupport" => %{"properties" => ["edit", "command"]}}, "completion" => %{"completionItem" => %{"deprecatedSupport" => true, "documentationFormat" => ["markdown", "plaintext"], "insertReplaceSupport" => true, "insertTextModeSupport" => %{"valueSet" => [1, 2]}, "resolveAdditionalTextEditsSupport" => true, "resolveSupport" => %{"properties" => ["documentation", "detail", "additionalTextEdits", "command"]}, "snippetSupport" => true}, "contextSupport" => true, "dynamicRegistration" => true}, "declaration" => %{"dynamicRegistration" => true, "linkSupport" => true}, "definition" => %{"dynamicRegistration" => true, "linkSupport" => true}, "documentLink" => %{"dynamicRegistration" => true, "tooltipSupport" => true}, "documentSymbol" => %{"hierarchicalDocumentSymbolSupport" => true, "symbolKind" => %{"valueSet" => [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]}}, "formatting" => %{"dynamicRegistration" => true}, "hover" => %{"contentFormat" => ["markdown", "plaintext"], "dynamicRegistration" => true}, "implementation" => %{"dynamicRegistration" => true, "linkSupport" => true}, "linkedEditingRange" => %{"dynamicRegistration" => true}, "onTypeFormatting" => %{"dynamicRegistration" => true}, "publishDiagnostics" => %{"relatedInformation" => true, "tagSupport" => %{"valueSet" => [1, 2]}, "versionSupport" => true}, "rangeFormatting" => %{"dynamicRegistration" => true}, "references" => %{"dynamicRegistration" => true}, "rename" => %{"dynamicRegistration" => true, "prepareSupport" => true}, "selectionRange" => %{"dynamicRegistration" => true}, "signatureHelp" => %{"dynamicRegistration" => true, "signatureInformation" => %{"parameterInformation" => %{"labelOffsetSupport" => true}}}, "synchronization" => %{"didSave" => true, "willSave" => true, "willSaveWaitUntil" => true}, "typeDefinition" => %{"dynamicRegistration" => true, "linkSupport" => true}, "typeHierarchy" => %{"dynamicRegistration" => true}}, "window" => %{"showDocument" => %{"support" => true}, "workDoneProgress" => true}, "workspace" => %{"applyEdit" => true, "codeLens" => %{"refreshSupport" => true}, "configuration" => true, "didChangeWatchedFiles" => %{"dynamicRegistration" => true}, "executeCommand" => %{"dynamicRegistration" => false}, "fileOperations" => %{"didCreate" => false, "didDelete" => false, "didRename" => true, "willCreate" => false, "willDelete" => false, "willRename" => true}, "symbol" => %{"symbolKind" => %{"valueSet" => [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]}}, "workspaceEdit" => %{"documentChanges" => true, "resourceOperations" => ["create", "rename", "delete"]}, "workspaceFolders" => true}}, "clientInfo" => %{"name" => "emacs", "version" => "GNU Emacs 29.1 (build 1, aarch64-apple-darwin22.6.0, NS appkit-2299.70 Version 13.5.2 (Build 22G91))\n of 2023-10-22"}, "initializationOptions" => nil, "processId" => 76120, "rootPath" => "/Users/johngoff/Repos/ci", "rootUri" => "file:///Users/johngoff/Repos/ci", "workDoneToken" => "1"}}

** (Protocol.UndefinedError) protocol Enumerable not implemented for nil of type Atom. This protocol is implemented for the following type(s): DBConnection.PrepareStream, DBConnection.Stream, Date.Range, Exqlite.Stream, File.Stream, Function, GenEvent.Stream, HashDict, HashSet, IO.Stream, Jason.OrderedObject, List, Map, MapSet, Range, Stream
    (elixir 1.15.7) lib/enum.ex:1: Enumerable.impl_for!/1
    (elixir 1.15.7) lib/enum.ex:166: Enumerable.reduce/3
    (elixir 1.15.7) lib/enum.ex:4387: Enum.reduce/3
    (next_ls 0.15.0) lib/next_ls.ex:643: NextLS.handle_notification/2
    (gen_lsp 0.6.2) lib/gen_lsp.ex:285: anonymous fn/4 in GenLSP.loop/3
    (gen_lsp 0.6.2) lib/gen_lsp.ex:308: GenLSP.attempt/3
    (stdlib 5.1.1) proc_lib.erl:241: :proc_lib.init_p_do_apply/3

```

- I've installed NextLS version 0.15.0 from the releases page.
- I tested on two projects, one with OTP 24.0.6 and Elixir 1.12.3, and one with OTP 25.2.1 and Elixir 1.14.3.
- I am using MacOS.
- My editor is Emacs 29.1 with lsp-mode, using the following config for lsp-mode: 
 ```
(after! lsp-mode
  (lsp-register-client
   (make-lsp-client :new-connection (lsp-stdio-connection '("nextls" "--stdio"))
                    :activation-fn (lsp-activate-on "elixir")
                    :server-id 'next-ls)))
```
  (I was intending on submitting the config to the docs once I confirmed it's working)

Happy to help debug, please let me know if there's anything I can test.

**Comment by mhanberg** [comment]
This looks like lsp-mode is sending nil for something in that payload that next ls expects to be a list, but its not clear whose fault that is.

i'll take a closer look at that payload

**Comment by mhanberg** [comment]
It looks like lsp-mode is declaring the "workspace folders" capabilities as "true", but is not sending up any workspace folders.

Just from glancing at some code on the lsp-mode repo, you probably want to try and add `:multi-root t` to your config

```lisp
(after! lsp-mode
 (lsp-register-client
  (make-lsp-client :new-connection (lsp-stdio-connection '("nextls" "--stdio"))
                   :multi-root t
                   :activation-fn (lsp-activate-on "elixir")
                   :server-id 'next-ls)))
```

**Comment by mhanberg** [comment]
If you get this working, would you mind contributing an `lsp-mode` section to the docs? (i see you are already familiar with contributing to the docs 😉)

**Comment by John-Goff** [comment]
Adding `:multi-root t` seems to have worked, no more crashes :tada: unfortunately, I am still getting this message when loading a project where the `mix.exs` is not in the project root:
```
[NextLS] Unexpected compiler response: {:error, %{message: "Could not find a Mix.Project, please ensure you are running Mix in a directory with a mix.exs file", __struct__: Mix.NoProjectError, __exception__: true, mix: true}}
```

**Comment by John-Goff** [comment]
If it helps, for this project I had to set the `elixirLs.projectDir` variable to the subdirectory containing the `mix.exs` file when using `elixir-ls`. Is there an equivalent setting for NextLS? From reading the docs, I was thinking that this multiple workspaces thing should handle that case.

**Comment by mhanberg** [comment]
That is what the workspace folders are for. 

I'll see how lsp-mode lets you specify the folders.

**Comment by John-Goff** [comment]
Can confirm it's working properly after I set up the folders properly. For context, I had a `root` directory which is the project root, and the `mix.exs` file and my elixir code live in the `server` directory. Not sure how this translates to other editors, but for the benefit of any future readers here's what I did with Doom Emacs.

I'm not sure if it's lsp-mode or projectile, but something picks up the project root as the workspace folder to add to lsp-mode. But it should actually be the directory which contains the `mix.exs`. So I called `M-x lsp-workspace-folders-remove`, selected `root` to be removed, then called `M-x lsp-workspace-folders-add`, and selected the `server` folder. After this, it works! Had to restart emacs to get it to pick up the right language server, but it works now.

Elixir LS provides a variable (`elixirLs.projectDir` in vscode, `lsp-elixir-project-dir` in emacs) that can be set to `"server"` to handle the case when the Elixir code is in a subdirectory of the workspace folder. As far as I could see, there is no equivalent setting for Next LS.
